File: packages/android/src/main/java/expo/modules/reactnativefx/FxLayoutObserver.kt
package expo.modules.reactnativefx

import android.graphics.Point
import android.graphics.Rect
import android.view.View
import androidx.core.graphics.Insets
import androidx.core.view.ViewCompat
import androidx.core.view.WindowInsetsCompat

/** Names a window edge for native travel measurement. */
internal enum class FxEdge {
  TOP,
  BOTTOM,
  LEFT,
  RIGHT,
}

/**
 * Captures the React-Native-assigned frame of a surface view and serves it to the
 * native animation driver by synchronous read. Reads layout, never writes it.
 *
 * React Native's mounting layer assigns the frame through `view.layout(...)`, which
 * dispatches layout-change listeners independently of the host's `onLayout` override —
 * the capture stays decoupled from the intermediate-container layout mechanic. The view
 * holds the listener on itself, so the observation lives exactly as long as the view and
 * survives the transient detach/reattach React Native performs on Android. Window-space
 * values are computed at read time because an ancestor scroll moves the view in the
 * window without a local layout event. All captures and reads happen on the UI thread.
 */
internal class FxLayoutObserver(private val view: View) : View.OnLayoutChangeListener {
  private val capturedFrame = Rect()

  init {
    view.addOnLayoutChangeListener(this)
    if (view.isLaidOut) {
      capturedFrame.set(view.left, view.top, view.right, view.bottom)
    }
  }

  override fun onLayoutChange(
    changedView: View,
    left: Int,
    top: Int,
    right: Int,
    bottom: Int,
    oldLeft: Int,
    oldTop: Int,
    oldRight: Int,
    oldBottom: Int,
  ) {
    capturedFrame.set(left, top, right, bottom)
  }

  /** Returns the captured post-layout frame in the parent's coordinate space, in px. */
  fun readFrameInParent(): Rect = Rect(capturedFrame)

  /** Returns the view's origin in window coordinates, in px. */
  fun readOriginInWindow(): Point {
    val location = IntArray(2)
    view.getLocationInWindow(location)
    return Point(location[0], location[1])
  }

  /**
   * Returns the distance the view travels to fully clear the named window edge — the
   * magnitude an edge-based enter or exit resolves natively.
   */
  fun readTravelDistance(edge: FxEdge): Float {
    val root = view.rootView ?: return 0f
    val location = IntArray(2)
    view.getLocationInWindow(location)
    return when (edge) {
      FxEdge.TOP -> (location[1] + view.height).toFloat()
      FxEdge.BOTTOM -> (root.height - location[1]).toFloat()
      FxEdge.LEFT -> (location[0] + view.width).toFloat()
      FxEdge.RIGHT -> (root.width - location[0]).toFloat()
    }
  }

  /** Returns the window's system-bar and cutout insets, or zero when unavailable. */
  fun readSafeAreaInsets(): Insets {
    val rootInsets = ViewCompat.getRootWindowInsets(view) ?: return Insets.NONE
    return rootInsets.getInsets(
      WindowInsetsCompat.Type.systemBars() or WindowInsetsCompat.Type.displayCutout()
    )
  }

  /** Stops observing layout changes. */
  fun detach() {
    view.removeOnLayoutChangeListener(this)
  }
}

File: packages/ios/FxLayoutObserver.swift
import UIKit

/// Names a window edge for native travel measurement.
internal enum FxEdge: String {
  case top
  case bottom
  case left
  case right
}

/// Captures the React-Native-assigned frame of a surface view and serves it to the
/// native animation driver by synchronous read. Reads layout, never writes it.
///
/// React Native applies layout by setting `center` and then `bounds` on the component
/// view, so observing `bounds` sees every applied frame — including origin-only moves —
/// with the origin already current when the observation fires. Window-space values are
/// computed at read time because an ancestor scroll moves the view in the window without
/// a local layout event. All captures and reads happen on the main thread.
internal final class FxLayoutObserver {
  /// The most recent React-Native-assigned frame: origin in the parent's space, in points.
  internal private(set) var capturedFrame: CGRect = .zero

  // Unowned because the view owns this observer, so the observer never outlives it.
  private unowned let view: UIView
  private var boundsObservation: NSKeyValueObservation?

  internal init(observing view: UIView) {
    self.view = view
    capturedFrame = view.frame
    boundsObservation = view.observe(\.bounds, options: [.new]) { [weak self] view, change in
      guard let newBounds = change.newValue else {
        return
      }
      self?.capturedFrame = CGRect(origin: view.frame.origin, size: newBounds.size)
    }
  }

  deinit {
    invalidate()
  }

  /// Returns the captured post-layout frame in the parent's coordinate space, in points.
  internal func readFrameInParent() -> CGRect {
    return capturedFrame
  }

  /// Returns the view's origin in window coordinates, or zero when detached.
  internal func readOriginInWindow() -> CGPoint {
    guard view.window != nil else {
      return .zero
    }
    return view.convert(view.bounds.origin, to: nil)
  }

  /// Returns the distance the view travels to fully clear the named window edge — the
  /// magnitude an edge-based enter or exit resolves natively. Zero when detached.
  internal func readTravelDistance(to edge: FxEdge) -> CGFloat {
    guard let window = view.window else {
      return 0
    }
    let frameInWindow = view.convert(view.bounds, to: nil)
    switch edge {
    case .top:
      return frameInWindow.maxY
    case .bottom:
      return window.bounds.height - frameInWindow.minY
    case .left:
      return frameInWindow.maxX
    case .right:
      return window.bounds.width - frameInWindow.minX
    }
  }

  /// Returns the safe-area insets currently applied to the view.
  internal func readSafeAreaInsets() -> UIEdgeInsets {
    return view.safeAreaInsets
  }

  /// Stops observing. Safe to call more than once.
  internal func invalidate() {
    boundsObservation?.invalidate()
    boundsObservation = nil
  }
}
